### ISSUE TITLE
OverdueStateApplicator

### ISSUE THREAD

**Issue by sbrossie** [killbill/killbill#535]
The logic inside `OverdueStateApplicator` is to rely on the current datetime from the `Clock` to set date of the following:

1. New `BlockingState` events
2. Cancelation date for subscription (`IMMEDIATE`) mode
3. New future notifications

The issue is that the `OverdueStateApplicator#apply` handler is called from many kind of notifications which can happen across different schedules: For instance there may be multiple payment retries schedules in parallel (whose own schedule is user driven through configuration), the overdue schedule (also driven through the overdue config with the `ReevaluationInterval`), and additional random notifications linked to user driven actions (insert charge, trigger payment).

For case `1.` and `2.`, in theory those should be triggered on time and so using `Clock.getUTCNow()` should be equivalent as computing the exact date at which transition needs to happen. Example using a 10 days `timeSinceEarliestUnpaidInvoiceEqualsOrExceeds` for all overdue states `OD1`, `OD2`:

* On `5/30` payment fails => Overdue System should set a notification for  `6/10` (10 days)
* On `6/10` notification triggers => Overdue System should set a notification for  `6/20` (10 days)

So as long as notifications are delivered on time this  should work, but there is a product question to answer in case notifications are late: Should we rely on `Clock.getUTCNow()`  or compute the date at which things should have happened (i think the later is the correct approach, especially in case we insert `BlockingState` events with `isBlockBilling` = `true` to ensure correct invoicing occurs.

For case `3.`, this is a bit different, and the current behavior is wrong: The code compute the `reevaluationInterval` and then based on `Clock.getUTCNow()`, it computes the next notification date (which then becomes arbitrary, because we enter this code at any time). To make things worst, there is another issue in [DefaultOverduePoster](https://github.com/killbill/killbill/blob/killbill-0.16.3/overdue/src/main/java/org/killbill/billing/overdue/notification/DefaultOverduePosterBase.java#L77) where we optimize the notifications that are inserted. Since we get notifications at random time we end up remove/adding random notifications. Instead we should ensure we have the notification at the right boundary times to ensure the overdue status transition at the right time. 

Note that case '3.' make the `TestOverdueIntegration` flaky because we sometimes miss the notification required to correctly transition state/insert `BlockingState` events. Example:

In the beatrix test `TestOverdueIntegration.testOverdueStages1`, we should see a notification at `7/20`, but we don't (and we also we see randomish notifications that were removed):

```
mysql> select queue_name, effective_date, processing_state from notifications_history where queue_name = 'overdue-service:overdue-check-queue' order by effective_date asc;
+-------------------------------------+---------------------+------------------+
| queue_name                          | effective_date      | processing_state |
+-------------------------------------+---------------------+------------------+
| overdue-service:overdue-check-queue | 2012-07-08 00:03:43 | PROCESSED        |
| overdue-service:overdue-check-queue | 2012-07-18 00:03:47 | REMOVED          |
| overdue-service:overdue-check-queue | 2012-07-18 00:03:47 | PROCESSED        |
| overdue-service:overdue-check-queue | 2012-07-28 00:03:49 | REMOVED          |
+-------------------------------------+---------------------+------------------+

```

Which leads to following failure:

```
2016-05-04T21:15:24.908+0000 [main] ERROR org.killbill.billing.api.TestApiListener - TestApiListener did not complete in 25000 ms, remaining events are BLOCK

java.lang.AssertionError: expected [true] but found [false]
Expected :true
Actual   :false
  <Click to see difference>
	at org.testng.Assert.fail(Assert.java:94)
	at org.testng.Assert.failNotEquals(Assert.java:494)
	at org.testng.Assert.assertTrue(Assert.java:42)
	at org.testng.Assert.assertTrue(Assert.java:52)
	at org.killbill.billing.api.TestApiListener.assertListenerStatus(TestApiListener.java:94)
	at org.killbill.billing.beatrix.integration.TestIntegrationBase.assertListenerStatus(TestIntegrationBase.java:269)
	at org.killbill.billing.beatrix.integration.TestIntegrationBase.doCallAndCheckForCompletion(TestIntegrationBase.java:743)
	at org.killbill.billing.beatrix.integration.TestIntegrationBase.addDaysAndCheckForCompletion(TestIntegrationBase.java:418)
	at org.killbill.billing.beatrix.integration.overdue.TestOverdueIntegration.testOverdueStages1(TestOverdueIntegration.java:150)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:84)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:714)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:901)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1231)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:127)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:111)
	at org.testng.TestRunner.privateRun(TestRunner.java:767)
	at org.testng.TestRunner.run(TestRunner.java:617)
	at org.testng.SuiteRunner.runTest(SuiteRunner.java:348)
	at org.testng.SuiteRunner.runSequentially(SuiteRunner.java:343)
	at org.testng.SuiteRunner.privateRun(SuiteRunner.java:305)
	at org.testng.SuiteRunner.run(SuiteRunner.java:254)
	at org.testng.SuiteRunnerWorker.runSuite(SuiteRunnerWorker.java:52)
	at org.testng.SuiteRunnerWorker.run(SuiteRunnerWorker.java:86)
	at org.testng.TestNG.runSuitesSequentially(TestNG.java:1224)
	at org.testng.TestNG.runSuitesLocally(TestNG.java:1149)
	at org.testng.TestNG.run(TestNG.java:1057)
	at org.testng.IDEARemoteTestNG.run(IDEARemoteTestNG.java:72)
	at org.testng.RemoteTestNGStarter.main(RemoteTestNGStarter.java:122)
```

By comparison, an iteration where the test succeeded shows:

```
select queue_name, effective_date, processing_state from notifications_history where queue_name = 'overdue-service:overdue-check-queue' order by effective_date asc;
+-----------------------------------------+---------------------+------------------+
| queue_name                              | effective_date      | processing_state |
+-----------------------------------------+---------------------+------------------+
| overdue-service:overdue-check-queue     | 2012-07-20 00:10:22 | AVAILABLE        |
| payment-service:retry                   | 2012-07-24 00:10:37 | AVAILABLE        |
| payment-service:retry                   | 2012-07-26 00:10:52 | AVAILABLE        |
| invoice-service:next-billing-date-queue | 2012-07-31 00:03:42 | AVAILABLE        |
+-----------------------------------------+---------------------+------------------+
```
